app/js/components/identifiers/addIdentifier.jsx

```diff
@@ -0,0 +1,231 @@
+        }));
+    }
+
+    callCreate() {
+        const keys = [];
+        const values = [];
+        const { addIdentifiers } = this.state;
+        Object.keys(addIdentifiers).filter((key) => {
+            if (addIdentifiers[key] !== '') {
+                keys.push(key);
+                values.push(addIdentifiers[key]);
+                return addIdentifiers[key];
+            }
+        });
+        const newAddIdentifiers = values.reduce((acc, cur, i) => (acc[keys[i]] = cur, acc), {});
+        if (this.state.addIdentifiers.preferred === true) {
+            apiCall(newAddIdentifiers,
+                'post', `patient/${this.props.stateData.patientUuid}/identifier`)
+                .then((response) => {
+                    apiCall({ preferred: false },
+                        'post',
+                        `patient/${this.props.stateData.patientUuid}/identifier/${this.props.stateData.prefferedIdentifierUuid}`,
+                    )
+                        .then((response) => {
+                            if (response.error) {
+                                this.setState({
+                                    addErrors: {
+                                        error: response.error.message,
+                                    },
+                                });
+                            } else {
+                                this.setState({ newState: false, addIdentifiers: {} });
+                                this.props.fetchIdentifiers();
+                            }
+                        })
+                        .catch(error => toastr.error(error));
+                })
+                .catch(error => toastr.error(error));
+        } else {
+            apiCall(newAddIdentifiers, 'post',
+                `patient/${this.props.stateData.patientUuid}/identifier`)
+                .then((response) => {
+                    if (response.error) {
+                        this.setState({ addErrors: { error: response.error.message } });
+                    } else {
+                        this.setState({ newState: false, addIdentifiers: {} });
+                        this.props.fetchIdentifiers();
+                    }
+                })
+                .catch(error => toastr.error(error));
+        }
+    }
+
+    callCancelPromise() {
+        return (new Promise((resolve, reject) => {
+            resolve(this.setState({
+                newState: false,
+            }, () => { this.props.callCancel; }));
+        },
+        ));
+    }
+
+    render() {
+        let editErrorClass = '';
+        let addErrorClass = '';
+        let editError = '';
+        let addError = '';
+        if (this.state.editErrors.error.length > 0) {
+            editErrorClass = 'has-error';
+            editError = this.state.editErrors.error;
+        }
+        if (this.state.addErrors.error.length > 0) {
+            addErrorClass = 'has-error';
+            addError = this.state.addErrors.error;
```

Comment:
Can you distructure or give this value `this.state.editErrors.error.length` a variable name since you used it 4 times? DRY

Reply:
done